Simplify calendar construction for Invariant mode

marek-safar opened the pull request:
Allows illinker to save more in invariant globalization mode

Not very accurate data for browser-wasm hello world as they will change as we progress on size trimming
|Mode  | SPC Size |
|-|-|
|Default| 1449KB | 
|Invariant| 1396KB |

marek-safar committed:
Simplify calendar construction for Invariant mode

Allows illinker to save about 50k in invariant globalization mode

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormatInfo.cs:
```diff
@@ -407,6 +407,14 @@
 
+                if (GlobalizationMode.Invariant)
+                {
+                    if (value.ID == calendar.ID)
```
>calendar [](start = 36, length = 8)

calendar can be null here

tarekgh reviewed (approved):
I added one comment you need to fix. other than that LGTM.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:
```diff
@@ -324,6 +324,11 @@ private void InitializeAbbreviatedEraNames(string localeName, CalendarId calenda
+            if (GlobalizationMode.Invariant)
+            {
+                return CalendarData.Invariant;
+            }
```
Do we already return invariant data for all calendarIds, and this is just doing the check earlier/differently to help the linker trim better?

Or is this actually changing behavior when GlobalizationMode.Invariant is set?

tarekgh replied:
In Invariant mode, we always use the Invariant culture when requesting any culture. so, I believe we are not changing any behavior for the invariant mode here.

marek-safar replied:
There is no behaviour change, PR only adds few explicit checks so less code is executed under GlobalizationMode.Invariant with the same result

marek-safar committed:
Fix nullable warning

marek-safar committed:
Add simple negative Calendar set test

marek-safar committed:
Retain setting of invariant calendar in non-invariant mode

jkotas commented:
```
    System.Globalization.Tests.InvariantModeTests.TestCultureData(cultureName: "en-US") [FAIL]
      System.ArgumentOutOfRangeException : Not a valid calendar for the given culture. (Parameter 'value')
      Actual value was System.Globalization.GregorianCalendar.
      Stack Trace:
        /_/src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormatInfo.cs(415,0): at System.Globalization.DateTimeFormatInfo.set_Calendar(Calendar value)
        /_/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs(731,0): at System.Globalization.CultureInfo.get_DateTimeFormat()
        /_/src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs(510,0): at System.Globalization.Tests.InvariantModeTests.TestCultureData(String cultureName)
```

marek-safar committed:
Fix initialization of DateTimeFormat for Invariant CultureInfo

to initialize  Calendar values only once.

Before CultureInfo::DateTimeFormat and DateTimeFormatInfo::InvariantInfo
would return different value under Invarint mode because of

```
at System.Globalization.DateTimeFormatInfo.set_Calendar(Calendar value)
at System.Globalization.DateTimeFormatInfo..ctor(CultureData cultureData, Calendar cal)
at System.Globalization.DateTimeFormatInfo..ctor()
at System.Globalization.DateTimeFormatInfo.get_InvariantInfo()
at System.Globalization.CultureInfo.get_DateTimeFormat()
at System.Globalization.CultureInfo.GetFormat(Type formatType)
at System.Globalization.DateTimeFormatInfo.get_CurrentInfo()
```

marek-safar committed:
Clarify existing assert

marek-safar committed:
Simplify dependencies when reading current era value

marek-safar committed:
Review feedback on the assert placement

safern reviewed (approved):
Other than one question, LGTM. @tarekgh could you take a look as well?

marek-safar committed:
Fix setting Calendar to same type in InvariantMode

marek-safar committed:
Fix odd behaviour where setting calendar to same value resets user values

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormatInfo.cs:
```diff
@@ -407,6 +407,17 @@ public Calendar Calendar
+                {
+                    if (value.ID == CalendarId.GREGORIAN)
+                    {
+                        calendar = value;
```
Fixed.

tarekgh reviewed (approved):
LGTM.

safern commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:
```diff
@@ -1671,7 +1671,7 @@ internal CalendarId[] CalendarIds
             get
             {
-                if (_waCalendars == null)
+                if (_waCalendars == null && !GlobalizationMode.Invariant)
```
Is this needed? based on this https://github.com/dotnet/runtime/pull/37527/files/48393f74f45999eb53ce79c7cfda61d7c30fb691..eee299b30be940db2a9857a70aab76de6a5eb55d#diff-235473f71512930b24ea62cbe21024f6R630 `_waCalendars` shouldn't be null for `Invariant` right?

marek-safar replied:
Yes but illinker cannot statically detect that

safern replied:
Ah I see, this is just to help the linker trim this code... ok cool.
